File: 018_four_sum/Solution.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

public class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        int arrLen = nums.length;
        List<List<Integer>> results = new ArrayList<>();
        if (nums == null || arrLen < 4) return results;

        Arrays.sort(nums);
        if (isExceeded(nums, target, 0, arrLen - 1)) return results;

        int L, R;

        for (int i = 0; i < arrLen - 3; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < arrLen - 2; j++) {
                if (j > i + 1 && nums[j] == nums[j - 1]) continue;
                L = j + 1;
                R = arrLen - 1;
                if (isExceeded(nums, target, i, R)) return results;

                while (L < R) {
                    int tmpSum = nums[i] + nums[j] + nums[L] + nums[R];
                    if (tmpSum == target) {
                        results.add(new ArrayList<>(Arrays.asList(nums[i], nums[j], nums[L], nums[R])));
                        while (L < R && nums[L] == nums[L + 1]) L++;
                        while (L < R && nums[R] == nums[R - 1]) R--;
                        L++; R--;
                    } else if (tmpSum > target) {
                        R--;
                    } else {
                        L++;
                    }
                }
            }
        }
        return results;
    }

    public boolean isExceeded(int[] nums, int target, int L, int R) {
        int minSum = nums[L] + nums[L + 1] + nums[L + 2] + nums[L + 3];
        int maxSum = nums[R] + nums[R - 1] + nums[R - 2] + nums[R - 3];
        return target > maxSum || target < minSum;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] nums = new int[]{1, 0, -1, 0, -2, 2};
        int target = 0;
        List<List<Integer>> results = solution.fourSum(nums, target);
        for (List<Integer> result : results) {
            for (int ele : result) {
                System.out.print(ele + " ");
            }
            System.out.println();
        }
    }
}

File: 018_four_sum/Solution.py
class Solution:
    def fourSum(self, nums: list, target: int) -> list:
        results = []
        arr_len = len(nums)
        if not nums or arr_len < 4: return results

        nums.sort()
        if self.is_exceeded(nums, target, 0, arr_len - 1): return results

        for i in range(arr_len - 3):
            if i > 0 and nums[i] == nums[i - 1]: continue
            for j in range(i + 1, arr_len - 2):
                if (j > i + 1 and nums[j] == nums[j - 1]): continue;
                l = j + 1
                r = arr_len - 1

                while l < r:
                    tmp_sum = nums[i] + nums[j] + nums[l] + nums[r]
                    if tmp_sum == target:
                        results.append([nums[i], nums[j], nums[l], nums[r]])
                        while l < r and nums[l] == nums[l + 1]: l += 1
                        while l < r and nums[r] == nums[r - 1]: r -= 1
                        l += 1
                        r -= 1
                        if self.is_exceeded(nums, target, i, r): continue
                    elif tmp_sum > target: r -= 1
                    else: l += 1

        return results

    def is_exceeded(self, nums: list, target: int, l: int, r: int) -> bool:
        min_sum = nums[l] + nums[l + 1] + nums[l + 2] + nums[l + 3]
        max_sum = nums[r] + nums[r - 1] + nums[r - 2] + nums[r - 3]
        return target > max_sum or target < min_sum


if __name__ == '__main__':
    nums = [1, 0, -1, 0, -2, 2]
    target = 0
    solution = Solution()
    print(solution.fourSum(nums, target))
